ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -1088,6 +1118,35 @@ void addChunkId(BlobId chunkBlobId, int chunkIndex) {
       indexToChunkIds.put(chunkIndex, chunkBlobId);
     }
 
+    /**
+     * Call {@link NotificationSystem#onBlobCreated(String, BlobProperties, byte[], NotificationBlobType)} for this
+     * chunk, unless it is the first chunk, in which case it might be an entire simple blob. In that case, save
+     * the {@link BlobProperties} from the first chunk.
+     * @param chunk the {@link PutChunk} created.
+     */
+    void maybeNotifyForChunkCreation(PutChunk chunk) {
+      if (chunk.chunkIndex == 0) {
+        firstChunkIdAndProperties = new Pair<>(chunk.chunkBlobId, chunk.chunkBlobProperties);
+      } else {
+        notificationSystem.onBlobCreated(chunk.chunkBlobId.getID(), chunk.chunkBlobProperties, userMetadata,
+            NotificationBlobType.DataChunk);
+      }
+    }
+
+    /**
+     * Notify for the creation of the first chunk if this was found to be a composite blob. To be called after the
+     * overall operation is completed.
+     */
+    void maybeNotifyForFirstChunkCreation() {
+      if (firstChunkIdAndProperties != null) {
+        String chunkId = firstChunkIdAndProperties.getFirst().getID();
+        BlobProperties chunkProperties = firstChunkIdAndProperties.getSecond();
+        if (!chunkId.equals(getBlobIdString())) {
```

Comment:
And this could simply omit this check - as with the previous change this method will be called if and only if the first chunk notification has to be done.